### PR TITLE
feat: superposition dynamic fields

### LINKED ISSUE #1518: Superposition in web-sdk
https://github.com/juspay/hyperswitch-sdk-cloud/issues/209#issue-4010170703

### PR DESCRIPTION
**Pull request by Shivam25092001**
## Type of Change

- [ ] Bugfix
- [X] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

- `DynamicFields.res` — Rewritten to use `useConfigurationService` → `getSuperpositionFinalFields`; RFF `<Form>` wraps all fields; billing split is path-based; card fields filtered at source to avoid double-render with CardPayment
- `DynamicFields/` — New per-field components: `CardHolderNameField`, `StateDropdownField`, `EmailField`, `CountryDropdownField`, `PhoneField`, etc.
- `DateOfBirth.res` — Refactored as RFF-wired DateField; this component was only getting used by `DynamicFields`
- `package.json` — Added `react-final-form`.
- `ReactFinalForm` owns all field state, validation, and initial values.

### Architectural Decisions
- No backward compatibility with the old `paymentMethodsFields` enum pipeline. `setRequiredFieldsBody` contract with all consumers is unchanged — they still receive `Dict.t<JSON.t>` keyed by API paths.
- RFF owns all field state, validation, and initial values. `useFormStateHandler` syncs `formProps.values` → `setRequiredFieldsBody` and `formProps.valid` → `areRequiredFieldsValid`.

### Why include RFF change in this PR?
- Using react-final-forms, the I/O based on `fieldConfigs` from superposition becomes very readable and easier to manage.
- Same change-radius.

### References
- [fieldConfig — key reference](https://docs.google.com/document/d/1RAQGhQLB391d8vXS3VzJ8BJOQLmeaV-0csDQnUM33vo/edit?usp=sharing) : This doc details about different fieldConfig keys.
- [Superposition HLD](https://docs.google.com/document/d/1js2JHCPXm6LHhpd_Hyt6pM81pSY7IVHbMsvVOnLuKdM/edit?tab=t.drlrmr5q5sl)

### Base PRs
- **[Shared code]** https://github.com/juspay/hyperswitch-sdk-utils/pull/57
- **[sdk configs api integration]** https://github.com/juspay/hyperswitch-web/pull/1577

### Stacked PRs
- Validation fixes and locales for error texts: https://github.com/juspay/hyperswitch-web/pull/1609.
- **[Shared code]** https://github.com/juspay/hyperswitch-sdk-utils/pull/59
- Intent Data collection from sdk-configs api: https://github.com/juspay/hyperswitch-web/pull/1602

### SCREEN RECORDING
https://github.com/user-attachments/assets/a6af807c-1cca-47e8-aad3-45ec137b3a28

### TESTING MATRIX
- Klarna - works
- Card - works
- AliPay - works
- Classic (cash2code) - works
- Evoucher - works
- Duit now (QR flow) - works
- Sepa debit - works
- Affirm - sandbox itself seems to be broken, superposition fix needed
- Eps - superposition fix needed
- Sepa bank transfer - works, but superposition changes needed
- Mifinity - Superposition changes required, works
- Online Banking FPX - Superposition changes required Bank selection]

## Issue
Fixes #1518 

## Checklist

- [X] I ran `npm run re:build`
- [X] I reviewed submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Comment by XyneSpaces** [comment]
## Code Review: Feat superposition dynamic fields clean

### Classification
**Feature:** Superposition Dynamic Fields with React Final Form integration  
**Risk Level:** High (form state management library addition, payment-critical code paths)  
**Files Changed:** 21 files (+1976/-1085 lines)

---

### Architecture Overview

This PR introduces **React Final Form (RFF)** for form state management in the dynamic fields system, significantly refactoring how form inputs handle state, validation, and submission.

| Area | Change |
|------|--------|
| Dependencies | Adds `react-final-form@^7.0.0` |
| Component Structure | Extracts 8 new field components from monolithic `DynamicFields.res` |
| State Management | Migrates from Recoil-only to RFF form state + Recoil for persistence |
| Validation | Centralized field validators in `DynamicFieldsUtils.res` |

---

### Tier 1 — Payment Correctness & Safety

#### ✅ No Silent-Null onFailure Patterns
Checked new field components — no error-swallowing patterns detected in the added/modified files.

#### ✅ No Card Data in Logs
No `console.log` statements with card-related fields found in diff.

#### ⚠️ Form State Boundary Risk in `LoaderController.res`
+230/-12 lines in the SDK entry point. Need to verify:
1. Form initialization doesn't lose payment intent state during recovery
2. Browser refresh behavior preserved for saved payment methods

#### ⚠️ Dependency Addition — React Final Form
Adding a new form state library requires scrutiny:
- Bundle size increase: `final-form@5.0.1` (~4KB gzipped) + `react-final-form@7.0.0` (~3KB gzipped)
- Peer dependency on `@babel/runtime`
- Form state lifecycle must align with SDK's mount/unmount patterns

**Action needed:** Document bundle size impact and verify RFF unmounts properly when SDK is destroyed.

---

### Tier 2 — Parity, Routing, Release Hygiene

#### ✅ v1/v2 Parity
Dynamic fields changes are in shared components — should apply to both v1 and v2.

#### ⚠️ Submodule Coordination Required
`DynamicFieldsUtils.res` changes affect field validation logic. Verify if `hyperswitch-client-core` has equivalent dynamic field logic needing updates.

#### ⚠️ Empty PR Body
The PR description is blank. Please add:
- What superposition dynamic fields are
- Why React Final Form was chosen over alternatives
- Migration plan for existing integrations

---

### Tier 3 — Style & Implementation Details

#### 💡 Good Code Organization
Extracting field components to `DynamicFields/` subdirectory is a maintainability win with clear single-responsibility boundaries.

#### 💡 Type Safety Improvement in `DynamicFieldsUtils.res`
Using `LocaleStringTypes.toValidationLocale` instead of `Obj.magic` eliminates runtime coercion risks.

#### ⚠️ Large Change in `PaymentUtils.res`
+366/-75 lines affecting payment payload construction. Needs verification that confirm payload structure matches backend expectations.

---

### Test Recommendations

1. **Matrix testing** — All payment methods using dynamic fields (cards with billing, BNPL, bank transfers)
2. **Edge cases** — Form reset after failed payment, prefilled customer data, browser autofill
3. **Bundle analysis** — Verify RFF addition doesn't exceed size budgets

---

### Summary

| Item | Status |
|------|--------|
| Silent-null patterns | ✅ None found |
| Card data exposure | ✅ None found |
| 3DS impact | ✅ N/A |
| Bundle size documented | ⚠️ Needs documentation |
| PR description | ⚠️ Empty - needs content |
| Client-core coordination | ⚠️ Verify needed |
| `LoaderController` risk | ⚠️ Careful testing required |

**Verdict:** ⚠️ **Request Changes** — This substantial refactoring improves organization but needs more documentation, bundle analysis, and cross-repo verification before merge.

**Review by XyneSpaces — COMMENTED** [review]
## Review Summary

**Classification:** Major refactoring — Migrates DynamicFields from Recoil-based state to React Final Form (RFF)
**Risk Level:** Medium (large-scale form architecture change, ~2000 lines modified)
**Scope:** Dynamic field rendering system, form validation, state management

---

### Findings Overview

| Severity | Count | Issues |
|----------|-------|--------|
| :rotating_light: Critical | 0 | None identified |
| :warning: Warning | 2 | Empty PR description, missing test checklist items |
| :bulb: Suggestion | 2 | Bundle size impact, error boundary consideration |

---

### :warning: Warnings

**1. Incomplete PR Description**

The PR body has empty sections for Description and Testing. For a change of this magnitude (+187/-783 in DynamicFields.res alone), the description should document:
- Migration rationale from Recoil to RFF
- Backward compatibility plan for existing field types
- Risk assessment for form validation edge cases

**2. Checklist Items Not Completed**

The PR checklist shows these items unchecked:
- [ ] I ran `npm run re:build`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible

Please confirm `npm run re:build` passes and code has been reviewed. Unit tests for the new field components would strengthen confidence in this refactor.

---

### :bulb: Suggestions

**1. Monitor bundle size impact of react-final-form**

The new dependency `react-final-form` (v7.0.0) brings in `final-form` (v5.0.1) as a peer dependency (~20KB minified). Verify this doesn't materially impact the SDK bundle size budget.

**2. Consider error boundary for field component crashes**

Individual field components now encapsulate more logic. Consider wrapping `<DynamicFieldInput.makeRow />` calls in an error boundary to prevent a single field misconfiguration from crashing the entire payment form.

---

### Tier 1 Verification Results

| Check | Status | Notes |
|-------|--------|-------|
| Silent-null onFailure | ✅ Pass | No new error-swallowing patterns in diff |
| Bare silent catch | ✅ Pass | No catch blocks that collapse to null |
| Card data in logs | ✅ Pass | No console logging of sensitive fields |
| v1/v2 parity | ℹ️ N/A | Appears to be form-layer only, no PaymentBody changes |
| Endpoint routing | ✅ Pass | No new URL literals |

---

### Tier 2 Verification Results

| Check | Status | Notes |
|-------|--------|-------|
| Submodule pointer moves | ✅ Pass | No sdk-utils version changes |
| Promote to shared-utils | ✅ Pass | New components are web-specific (RFF-dependent) |
| Version sync | ℹ️ N/A | No version bump in package.json |

---

### Positive Notes

- ✅ Clean separation of concerns with new `DynamicFields/` subdirectory
- ✅ Good use of RFF `useField` hook pattern across all new components
- ✅ Proper validation integration via `DynamicFieldsUtils.resolveValidator`
- ✅ `GenericInputField` uses RFF `<Field>` render prop pattern correctly
- ✅ `DateOfBirth` refactored to integrate with RFF validation
- ✅ `LoaderController` properly guards sdkConfigsValue update behind error check

---

### Pre-existing Issues (unchanged)

None introduced by this PR.

---

**Overall Assessment:**
This is a well-structured migration to React Final Form that should improve form state management consistency. The risk is moderate due to the large footprint. Completing the PR description and checklist would help reviewers understand the full scope.

**Verdict:** :speech_balloon: **Comment** — Clean refactor. Please complete PR description and checklist before merge.

**Review by XyneSpaces — COMMENTED** [review]
## Review Summary

**Verdict:** :arrows_counterclockwise: Request Changes

:rotating_light: 1 critical · :warning: 2 warnings · :bulb: 2 suggestions

This is a substantial architectural change introducing React Final Form across the dynamic fields system. The dependency addition and state management changes warrant careful review of error handling and backward compatibility.

**Classification:** core-flow-framework · secondary: sdk-ffi-codegen
**Risk:** high

File: package.json [hunk 1]
```diff
@@ -12,6 +12,7 @@
     "react": "^18.2.0",
```
Comment (XyneSpaces) [inline package.json:12]: :warning: New dependency `react-final-form@^7.0.0` - verify no duplicate `final-form` peer dependency in bundle.

File: package-lock.json [hunk 2]
```diff
@@ -17,6 +17,7 @@
         "react": "^18.2.0",
```
Comment (XyneSpaces) [inline package-lock.json:17]: :bulb: Ensure `@babel/runtime@7.29.2` addition doesn't conflict with existing babel runtime versions.

**Review by XyneSpaces — COMMENTED** [review]
:warning: **React Final Form dependency addition requires peer dependency verification**

`react-final-form` v7.0.0 pulls in `final-form` v5.0.1 as a peer. The RFF v7 line requires React 16.8+. Verify compatibility with the project's React 18.2.0 and ensure no duplicate `final-form` instances in the bundle.

**Fix**: Confirm `npm ls final-form` produces a single instance.

File: src/Components/DynamicFields/GenericDropdownField.res [hunk 3]
```diff
@@ -0,0 +1,24 @@
+open SuperpositionTypes
+
```
Comment (AbhishekChorotiya) [inline src/Components/DynamicFields/GenericDropdownField.res:2]: merge this component to GenericInputField and use switch on fieldType to render

File: src/Components/DynamicFields/CardHolderNameField.res [hunk 4]
```diff
@@ -0,0 +1,97 @@
+open SuperpositionTypes
+
+// "John Doe" -> ("John", "Doe")   "John" -> ("John", "")
+let splitName = (value: string): (string, string) =>
```
Comment (aritro2002) [inline src/Components/DynamicFields/CardHolderNameField.res]: check if we can use this function `getFirstAndLastNameFromFullName` instead of creating `splitName`

File: src/Components/DynamicFields/StateDropdownField.res [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  let firstStateName = stateDisplayNames->Array.get(0)->Option.getOr("")
+  let effectiveDisplayName = storedDisplayName !== "" ? storedDisplayName : firstStateName
+  let effectiveCode = Utils.getStateCodeFromStateName(effectiveDisplayName, countryIso)
+
+  React.useEffect(() => {
+    if hasStates && storedCode !== effectiveCode {
+      stateField.input.onChange(effectiveCode)
+    }
+    None
+  }, [effectiveCode])
```
Comment (aritro2002) [inline src/Components/DynamicFields/StateDropdownField.res]: Incomplete useEffect Dependency Array
```suggestion
  }, [hasStates, storedCode, effectiveCode])
```

File: src/Components/DynamicFields/CardHolderNameField.res [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+      lastNameFieldConfig.confirmRequestWritePath,
+      ~config={validate: lastValidator},
+    )
+
+    let inputRef = React.useRef(Nullable.null)
+    let (name, setName) = React.useState(() => {
+      let first = firstProps.input.value->Option.getOr("")
+      let last = lastProps.input.value->Option.getOr("")
+      [first, last]->Array.filter(val => val !== "")->Array.join(" ")
+    })
```
Comment (aritro2002) [inline src/Components/DynamicFields/CardHolderNameField.res:31]: This only runs ONCE at mount. If firstProps/lastProps aren't populated yet (e.g. async form init, or values come in after mount), this will be "" forever.

File: src/Components/DynamicFields/CardHolderNameField.res (hunk 4 above)
Comment (aritro2002) [reply src/Components/DynamicFields/CardHolderNameField.res]: "John  Doe" (two spaces) would produce ("John", " Doe") with a leading space in the last name. This case will be missed by `splitName`

File: src/LoaderController.res [hunk 7]
```diff
@@ -626,6 +627,9 @@ let make = (~children, ~paymentMode, ~setIntegrateErrorError, ~logger, ~initTime
           | _ => ()
           }
           setSdkConfigs(_ => updatedState)
+          if !isSdkConfigsError {
```
Comment (aritro2002) [inline src/LoaderController.res:630]: why is there need of `isSdkConfigsError`, if no response comes, we can set the default value, right?

File: src/Components/DynamicFields/PhoneField.res [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  let {label, placeholder} = DynamicFieldsUtils.resolveFieldTexts(
+    ~field=fieldConfig,
+    ~localeObject=localeString,
+  )
+  let maxLength = fieldConfig.maxInputLength
+  let autocomplete = fieldConfig.htmlAutocompleteAttribute->Option.getOr("tel-national")
+
+  let validate = DynamicFieldsUtils.resolveValidator(~field=fieldConfig, ~localeObject=localeString)
+
+  let field = ReactFinalForm.useField(path, ~config={validate: validate})
```
Comment (AbhishekChorotiya) [inline src/Components/DynamicFields/PhoneField.res:17]: use  <ReactFinalForm.Field> instead of useField

File: src/Components/DynamicFieldInput.res [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+  | CardHolderName =>
+    let allCardHolderNameFields = switch globalCardHolderNameFields {
+    | Some(nameFields) => nameFields
+    | None => []
+    }
+    let firstCardHolderNamePath =
+      allCardHolderNameFields->Array.get(0)->Option.map(field => field.confirmRequestWritePath)
+    <RenderIf condition={firstCardHolderNamePath === Some(field.confirmRequestWritePath)}>
+      <CardHolderNameField fields=allCardHolderNameFields />
+    </RenderIf>
```
Comment (AbhishekChorotiya) [inline src/Components/DynamicFieldInput.res:56]: render a single component here and move the rendering logic inside the react component 
```
| CardHolderName => <CardHolderNameField ... />
```

Comment (AbhishekChorotiya) [reply src/Components/DynamicFieldInput.res:56]: do same for all other cases

File: src/Components/DynamicFields/EmailField.res [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    let invalid = primaryField.meta.invalid
+    let isValid = touched ? Some(!invalid) : None
+    let errorString = touched && invalid ? primaryField.meta.error->Option.getOr("") : ""
+
+    <PaymentInputField
+      fieldName={label}
+      value
+      onChange={ev => {
+        let val = ReactEvent.Form.target(ev)["value"]
+        fields->Array.forEach(field => form.change(field.confirmRequestWritePath, val))
```
Comment (AbhishekChorotiya) [inline src/Components/DynamicFields/EmailField.res:38]: primaryField is not mapped with onChange, update primaryField value here so that ReactFinalForms can properly track field state

Comment (AbhishekChorotiya) [reply src/Components/DynamicFields/EmailField.res:38]:
```
 <PaymentInputField
        fieldName={label}
        value
        onChange={ev => {
          let val: string = ReactEvent.Form.target(ev)["value"]
          // Drive the registered field through its own handler...
          primaryField.input.onChange(val)

          secondaryPaths->Array.forEach(path => form.change(path, val))
        }}
        onBlur={_ev => primaryField.input.onBlur()}
        isValid
        errorString
        placeholder
        inputRef={fieldRef}
        autocomplete
      />
```

File: src/Components/DynamicFields/PhoneCountryCodeDropdownField.res [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+  let {label} = DynamicFieldsUtils.resolveFieldTexts(~field=fieldConfig, ~localeObject=localeString)
+  let validate = DynamicFieldsUtils.resolveValidator(~field=fieldConfig, ~localeObject=localeString)
+
+  let countryAndCodeList =
+    phoneNumberJson
+    ->JSON.Decode.object
+    ->Option.getOr(Dict.make())
+    ->getArray("countries")
+
+  let phoneNumberCodeOptions: array<DropdownField.optionType> = countryAndCodeList->Array.reduce(
```
Comment (AbhishekChorotiya) [inline src/Components/DynamicFields/PhoneCountryCodeDropdownField.res]: this value is getting recomputed on every render, please use useMemo here

File: src/Components/DynamicFields/PhoneCountryCodeDropdownField.res [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+let getPhoneCode = val => val->String.split("#")->Array.get(1)->Option.getOr("")
+
+@react.component
+let make = (~fieldConfig: fieldConfig) => {
+  open Utils
+  let {config, localeString} = Recoil.useRecoilValueFromAtom(RecoilAtoms.configAtom)
+  let {label} = DynamicFieldsUtils.resolveFieldTexts(~field=fieldConfig, ~localeObject=localeString)
+  let validate = DynamicFieldsUtils.resolveValidator(~field=fieldConfig, ~localeObject=localeString)
+
+  let countryAndCodeList =
```
Comment (AbhishekChorotiya) [inline src/Components/DynamicFields/PhoneCountryCodeDropdownField.res]: useMemo()

File: src/Components/DynamicFields/PhoneCountryCodeDropdownField.res [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    phoneNumberJson
+    ->JSON.Decode.object
```
Comment (AbhishekChorotiya) [inline src/Components/DynamicFields/PhoneCountryCodeDropdownField.res]: Utils.getDictFromJson

File: src/Components/DynamicFieldInput.res [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+  ~allFields: array<fieldConfig>,
+  ~globalEmailFields: option<array<fieldConfig>>=?,
+  ~globalCardHolderNameFields: option<array<fieldConfig>>=?,
+) => {
+  switch field.fieldRenderType {
+  | CardNumber
+  | Cvc
+  | CardExpiryMonth
+  | CardExpiryYear
+  | CardNetwork => React.null
```
Comment (AbhishekChorotiya) [inline src/Components/DynamicFieldInput.res:38]: how card fields are getting rendered via RRF + superposition?

File: src/Utilities/DynamicFieldsUtils.res [hunk 15]
```diff
@@ -1,166 +1,350 @@
...
-let addBillingAddressIfUseBillingAddress = (
-  fieldsArr,
+let removeBillingDetailsIfUseBillingAddress = (
+  missingRequiredFields: array<SuperpositionTypes.fieldConfig>,
   billingAddress: PaymentType.billingAddress,
 ) => {
   if billingAddress.isUseBillingAddress {
-    fieldsArr->Array.concat(billingAddressFields)
+    missingRequiredFields->Array.filter(requiredField => {
+      !(requiredField.confirmRequestWritePath->String.startsWith("payment_method_data.billing."))
```
Comment (AbhishekChorotiya) [inline src/Utilities/DynamicFieldsUtils.res]: use billingPrifix variable instead of hardcoding the value

File: src/Utilities/DynamicFieldsUtils.res [hunk 16]
```diff
@@ -1,166 +1,350 @@
...
+    switch field.validationRegexPattern {
+    | Some(pattern) => [Validation.Generic(pattern)]
+    | None => []
+    }
+  | _ => []
   }
+
+  let maxLengthRule = [Validation.MaxLength(field.maxInputLength->Option.getOr(255))]
+
+  let rules = Array.concat(Array.concat(requiredRule, semanticRule), maxLengthRule)
```
Comment (AbhishekChorotiya) [inline src/Utilities/DynamicFieldsUtils.res]: let rules = [...requiredRule, ...semanticRule, ...maxLengthRule]

File: src/Components/DynamicFields.res [hunk 17]
```diff
@@ -32,834 +32,238 @@ let make = (
...
+    <RenderIf condition={!isSavedCardFlow && hasAnyField}>
+      <ReactFinalForm.Form
+        initialValues={Some(initialValues)}
+        onSubmit={_values => ()}
+        render={formProps => {
+          formRef.current = Some(formProps.form)
+
+          ReactFinalForm.useFormStateHandler(
+            ~onFormChange=values => {
+              // Flatten the nested form values so keys align correctly during merge using `mergeAndFlattenToTuples`.
```
Comment (AbhishekChorotiya) [inline src/Components/DynamicFields.res]: please correct the comment or remove

File: src/Components/DynamicFields/GenericDropdownField.res (hunk 3 above)
Comment (Shivam25092001) [reply src/Components/DynamicFields/GenericDropdownField.res:2]: GenericDropdown and GenericInput fields have different behaviors at the UI layer, so kept them as separate components.

Also, wanted to keep the switch logic at a single place at the parent component, i.e. at `DynamicFieldInput.res`'s function `renderSingleField`

File: src/Components/DynamicFields/PhoneField.res (hunk 8 above)
Comment (Shivam25092001) [reply src/Components/DynamicFields/PhoneField.res:17]: All other components are using `ReactFinalForm.useField`. Any specific reason you can point for using `use <ReactFinalForm.Field>` here?

File: src/Components/DynamicFieldInput.res (hunk 9 above)
Comment (Shivam25092001) [reply src/Components/DynamicFieldInput.res:56]: Wanted to keep the logic of RenderIf at the root itself, so that the component itself isn't mounted incase it's not to be rendered. 

Also, I feel it's more readable to show if a component is rendered at the root itself, rather than encapsulating it in the Component's internal render tree.

File: src/LoaderController.res (hunk 7 above)
Comment (Shivam25092001) [reply src/LoaderController.res:630]: Used for logging.

File: src/Components/DynamicFields/EmailField.res (hunk 10 above)
Comment (Shivam25092001) [reply src/Components/DynamicFields/EmailField.res:38]: image
`field.input.onChange` internally uses `form.change`

So, `fields->Array.forEach(field => form.change(field.confirmRequestWritePath, val))` already handles `primaryField.input.onChange`

File: src/Components/DynamicFieldInput.res (hunk 14 above)
Comment (Shivam25092001) [reply src/Components/DynamicFieldInput.res:38]: CardFields aren't impacted in this changes, we are using `CardPayment.res` to render those fields.


Dataflow:

- `setRequiredFieldsBody` is passed down as a prop to `<DynamicFields>`
- in `DynamicFields`, RFF's `useFormStateHandler` pushes the flattened form values into it 
- at confirm time, `CardPayment`'s `submitCallback` assembles everything
```
     cardBody ─► mergeAndFlattenToTuples(requiredFieldsBody)       [:403 / Utils.res:1798]
           • flattens cardBody, merges the RFF-sourced requiredFieldsBody dict on top
     └─► intent(~bodyArr=..., ~confirmParam, ...)  ── the confirm call   [:399-408]
```

File: src/Components/DynamicFields/CardHolderNameField.res (hunk 6 above)
Comment (Shivam25092001) [reply src/Components/DynamicFields/CardHolderNameField.res:31]: DynamicFeilds are only rendered once initialValues are already calculated. So, this component mounts only after intitalValues have already been populated in RFF, so this case won't be there.

File: src/Components/DynamicFields/CardHolderNameField.res (hunk 4 above)
Comment (Shivam25092001) [reply src/Components/DynamicFields/CardHolderNameField.res]: done

File: src/Components/DynamicFields/StateDropdownField.res (hunk 5 above)
Comment (Shivam25092001) [reply src/Components/DynamicFields/StateDropdownField.res]: done 👍

File: src/Components/DynamicFields/PhoneCountryCodeDropdownField.res (hunk 11 above)
Comment (Shivam25092001) [reply src/Components/DynamicFields/PhoneCountryCodeDropdownField.res]: done 👍

File: src/Components/DynamicFields/PhoneCountryCodeDropdownField.res (hunk 12 above)
Comment (Shivam25092001) [reply src/Components/DynamicFields/PhoneCountryCodeDropdownField.res]: done 👍

File: src/Components/DynamicFields/PhoneCountryCodeDropdownField.res (hunk 13 above)
Comment (Shivam25092001) [reply src/Components/DynamicFields/PhoneCountryCodeDropdownField.res]: done 👍

File: src/Utilities/DynamicFieldsUtils.res (hunk 15 above)
Comment (Shivam25092001) [reply src/Utilities/DynamicFieldsUtils.res]: done 👍

File: src/Utilities/DynamicFieldsUtils.res (hunk 16 above)
Comment (Shivam25092001) [reply src/Utilities/DynamicFieldsUtils.res]: done 👍

File: src/Components/DynamicFields.res (hunk 17 above)
Comment (Shivam25092001) [reply src/Components/DynamicFields.res]: done 👍

File: src/Components/DynamicFields/PhoneField.res (hunk 8 above)
Comment (AbhishekChorotiya) [reply src/Components/DynamicFields/PhoneField.res:17]: it was use inconstantly, please follow one pattern for field binding

File: src/Components/DynamicFields.res [hunk 18]
```diff
@@ -32,834 +32,237 @@ let make = (
...
+            ~onFormChange=values => {
+              // RFF stores values as nested objects; flatten to dot-notation keys so they align with the confirm-payload merge.
+              let flatValues = values->JSON.Encode.object->Utils.flattenObject(false)
+              setRequiredFieldsBody(_ => flatValues)
+            },
+            ~onValidationChange=isValid => {
+              setAreRequiredFieldsValid(_ => isValid)
+            },
+            ~formProps,
+          )
```
Comment (aritro2002) [inline src/Components/DynamicFields.res:218]: ReactFinalForm.useFormStateHandler(...) is invoked inside the render callback of <ReactFinalForm.Form>. If this is a React hook (strongly implied by the use prefix), it violates the Rules of Hooks (https://react.dev/warnings/invalid-hook-call-warning) — hooks must be called at the top level of a component function, not inside callbacks, conditions, or loops

File: src/Components/DynamicFields.res [hunk 19]
```diff
@@ -32,834 +32,237 @@ let make = (
...
-    paymentMethodTypesV2.requiredFields,
-    paymentMethodType,
-    creditPaymentMethodTypes.required_fields,
-    creditPaymentMethodTypesV2.requiredFields,
+    getSuperpositionFinalFields,
+    eligibleConnectors,
+    superpositionBaseContext,
+    requiredFieldsFromPMLFlat,
   ))
+  let initialValues = React.useMemo(() => superpositionInitialValues, [requiredFieldsFromPMLFlat])
```
Comment (aritro2002) [inline src/Components/DynamicFields.res:93]: the dependency array only includes requiredFieldsFromPMLFlat, but superpositionInitialValues also depends on eligibleConnectors and superpositionBaseContext. If those change without requiredFieldsFromPMLFlat changing, the memoized value will be stale.

File: src/Components/DynamicFields.res [hunk 20]
```diff
@@ -32,834 +32,237 @@ let make = (
...
+  let hasAnyField = missingRequiredFieldsFiltered->Array.length > 0
+  let setAreRequiredFieldsValid = Recoil.useSetRecoilState(areRequiredFieldsValid)
+
+  <>
+    <RenderIf condition={!isSavedCardFlow && hasAnyField}>
+      <ReactFinalForm.Form
+        initialValues={Some(initialValues)}
+        onSubmit={_values => ()}
+        render={formProps => {
+          formRef.current = Some(formProps.form)
```
Comment (aritro2002) [inline src/Components/DynamicFields.res:197]: formRef.current = Some(formProps.form) mutates a ref during the render phase. In React 18's concurrent rendering, this can run multiple times or be discarded, leading to the ref holding stale or incorrect values.

✅ Suggested fix — use useEffect
```
@react.component
let make = (~formProps, ~formRef) => {
  React.useEffect(() => {
    formRef.current = Some(formProps.form)
    None
  }, [formProps.form])
  ...
}
```

File: src/Components/DynamicFields.res (hunk 18 above)
Comment (Shivam25092001) [reply src/Components/DynamicFields.res:218]: We can move the logic to a separate component, it's just a linting flag. since the hooks isn't being called conditionally and is at top level of the render block, it itsn't an issue

File: src/Components/DynamicFields/PhoneField.res (hunk 8 above)
Comment (Shivam25092001) [reply src/Components/DynamicFields/PhoneField.res:17]: done

File: src/Components/DynamicFields.res (hunk 19 above)
Comment (Shivam25092001) [reply src/Components/DynamicFields.res:93]: Memoizing the initialValue that should only change on change of requiredFieldsFromPMLFlat, shouldn't be effected by change of other parameters

File: src/Components/DynamicFields.res [hunk 21]
```diff
@@ -32,834 +32,255 @@ let make = (
...
     ~paymentMethodListValue,
     ~paymentMethod,
     ~paymentMethodType,
   )
 
-  let paymentMethodTypesV2 = PaymentUtilsV2.usePaymentMethodTypeFromListV2(
-    ~paymentsListValueV2=paymentManagementListValue,
-    ~paymentMethod,
-    ~paymentMethodType,
-  )
```
Comment (sakksham7) [inline src/Components/DynamicFields.res:64]: We should not remove this Dynamic field logic handling right?
Wherever it's V2 related since this is getting used in Payment Methods management flow which has dynamic fields in Combined Payment methods List of Payment Methods session, and there will be no sdk-config involved in that flow

**Review by AbhishekChorotiya — APPROVED** [review]

**Review by aritro2002 — APPROVED** [review]
